reference/test/factorization/par_ilu_kernels.cpp

```diff
@@ -0,0 +1,426 @@
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/factorization/par_ilu.hpp>
+
+
+#include <memory>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/factorization/par_ilu_kernels.hpp"
+#include "core/test/utils/assertions.hpp"
+
+
+namespace {
+
+
+class DummyLinOp : public gko::EnableLinOp<DummyLinOp>,
+                   public gko::EnableCreateMethod<DummyLinOp> {
+public:
+    DummyLinOp(std::shared_ptr<const gko::Executor> exec,
+               gko::dim<2> size = gko::dim<2>{})
+        : EnableLinOp<DummyLinOp>(exec, size)
+    {}
+
+protected:
+    void apply_impl(const gko::LinOp *b, gko::LinOp *x) const override {}
+
+    void apply_impl(const gko::LinOp *alpha, const gko::LinOp *b,
+                    const gko::LinOp *beta, gko::LinOp *x) const override
+    {}
+};
+
+
+class ParIlu : public ::testing::Test {
+protected:
+    using value_type = gko::default_precision;
+    using index_type = gko::int32;
+    using Dense = gko::matrix::Dense<value_type>;
+    using Coo = gko::matrix::Coo<value_type, index_type>;
+    using Csr = gko::matrix::Csr<value_type, index_type>;
+    ParIlu()
+        : ref(gko::ReferenceExecutor::create()),
+          exec(std::static_pointer_cast<const gko::Executor>(ref)),
+          identity(gko::initialize<Dense>(
+              {{1., 0., 0.}, {0., 1., 0.}, {0., 0., 1.}}, exec)),
+          lower_triangular(gko::initialize<Dense>(
+              {{1., 0., 0.}, {1., 1., 0.}, {1., 1., 1.}}, exec)),
+          upper_triangular(gko::initialize<Dense>(
+              {{1., 1., 1.}, {0., 1., 1.}, {0., 0., 1.}}, exec)),
+          mtx_small(gko::initialize<Dense>(
+              {{4., 6., 8.}, {2., 2., 5.}, {1., 1., 1.}}, exec)),
+          mtx_csr_small(nullptr),
+          small_l_expected(gko::initialize<Dense>(
+              {{1., 0., 0.}, {0.5, 1., 0.}, {0.25, 0.5, 1.}}, exec)),
+          small_u_expected(gko::initialize<Dense>(
+              {{4., 6., 8.}, {0., -1., 1.}, {0., 0., -1.5}}, exec)),
+          mtx_big(gko::initialize<Dense>({{1., 1., 1., 0., 1., 3.},
+                                          {1., 2., 2., 0., 2., 0.},
+                                          {0., 2., 3., 3., 3., 5.},
+                                          {1., 0., 3., 4., 4., 4.},
+                                          {1., 2., 0., 4., 5., 6.},
+                                          {0., 2., 3., 4., 5., 8.}},
+                                         exec)),
+          big_l_expected(gko::initialize<Dense>({{1., 0., 0., 0., 0., 0.},
```

Comment:
I tried this out in MATLAB. When I multiply L and U I do not get back the matrix you have above. You can try it with this test file.
[test_ilu.txt](https://github.com/ginkgo-project/ginkgo/files/3275159/test_ilu.txt)

Reply:
Ah, I see. Sorry. My fault.